formats/html.ts

```diff
@@ -21,24 +73,340 @@ export class HtmlImporter extends FormatImporter {
+			return "failed";
+		}
+	}
+
+	async requestFile(url: URL) {
+		const data = (await fsPromises.readFile(nodeUrl.fileURLToPath(url.href))).buffer;
+		return { mime: await detectMime(url, data), data };
+	}
+
+	async requestHTTP(url: URL) {
+		url = new URL(url.href);
+		let response;
+		try {
+			url.protocol = "https:";
+			response = await requestUrl({
+				url: url.href,
+				method: "GET",
+				throw: true,
+			});
+		} catch (e) {
+			try {
+				url.protocol = "http:";
+				response = await requestUrl({
+					url: url.href,
+					method: "GET",
+					throw: true,
+				});
+			} catch (e2) {
+				console.error(e2);
+				throw e;
+			}
+		}
+		const { arrayBuffer: data } = response;
+		return { mime: response.headers["Content-Type"] || await detectMime(url, data), data };
+	}
 
-		for (let file of files) {
+	async writeAttachment(mdFile: TFile, filename: string, data: ArrayBufferLike) {
+		const { basename, extension } = splitFilename(sanitizeFileName(filename));
+		let error;
+		for (let retry = 0; retry < 5; ++retry) {
 			try {
-				let htmlContent = await file.readText();
-				let mdContent = htmlToMarkdown(htmlContent);
-				await this.saveAsMarkdownFile(folder, file.basename, mdContent);
-				results.total++;
+				//@ts-ignore
+				const path: string = await this.app.vault.getAvailablePathForAttachments(basename, extension, mdFile);
+				return await this.app.vault.createBinary(path, data);
 			} catch (e) {
-				console.error(e);
-				results.failed.push(file.toString());
+				// retry in case `path` is the same for multiple invocations of `writeAttachment`
+				error = e;
+				await sleep(1000 * Math.random());
 			}
 		}
+		throw error;
+	}
 
-		this.showResult(results);
+	async filterAttachment(response: Response) {
+		const { data, mime } = response;
+		return this.filterAttachmentSize(data) && this.filterAttachmentMime(mime) && await this.filterImageSize(response);
+	}
+
+	filterAttachmentSize(data: ArrayBufferLike) {
+		const { byteLength } = data;
+		return !this.attachmentSizeLimit || byteLength <= this.attachmentSizeLimit;
+	}
+
+	filterAttachmentMime(mime: string) {
+		return ["audio/", "image/", "video/"]
+			.some(prefix => mime.startsWith(prefix));
+	}
+
+	async filterImageSize(response: Response) {
+		const { mime, data } = response;
+		if (!this.minimumImageSize || !mime.startsWith("image/")) {
+			return true;
+		}
+		let size;
+		try {
+			size = await imageSize(new Blob([data], { type: mime }));
+		} catch {
+			return true;
+		}
+		const { height, width } = size;
+		return width >= this.minimumImageSize && height >= this.minimumImageSize;
+	}
+}
+
+interface Response {
+	mime: string;
+	data: ArrayBufferLike;
+}
+
+function getURLFilename(url: URL) {
+	return pathToFilename(normalizePath(decodeURI(url.pathname)));
+}
+
+async function detectMime(url: URL, data: ArrayBufferLike) {
+	return mime(splitFilename(getURLFilename(url)).extension) ||
+		((await fileType.fileTypeFromBuffer(data))?.mime ??
```

Comment:
Hmm I've been thinking about it - I think I am ok to forego binary based detection. We should prioritize it like this:
- Mime type from the response, if available, and if it converts to an extension correctly
- Extension from the URL if possible
- Skip the file otherwise.

Reply:
Okay, the binary-based detection is gone now.

Actually, with this idea and the idea above to use `DOMParser` to read the attachments, the entire MIME detection can be eliminated. Basically the steps are:
1. Use the `DOMParser` to read the attachments. Each attachment is labelled whether it is from a `<audio>`, `<img>`, or `<video>` HTML element.
2. (some unrelated intermediate steps later)
3. Check whether the downloaded file has a content type or file extension. Lookup or keep the extension if so. Otherwise, `.noext.mp3` (or any other audio file extension) is appended if the attachment is from an `<audio>`, `.noext.bmp` (or any other image file extension) for `<img>`, and `.noext.mp4` for `<video>` (or any other video file extension). It doesn't matter whether the file format actually matches the extension as the Markdown renderer only uses the extension to determine whether to use a `<audio>`, `<img>`, `<video>`, or no embed. The browser will figure out the actual file format by itself. This way, if we cannot detect the extension, the file can still be displayed. `.noext` here is used to label that they originally have no extensions to the user.

Reply:
That's amazing. One suggestion I had is to default to png or jpg instead of bmp, since those are much more commonly used these days.

Reply:
> That's amazing. One suggestion I had is to default to png or jpg instead of bmp, since those are much more commonly used these days.

Replaced wih `png` now.